src/security/token/TKTokenCryptoEngine.m

```diff
@@ -48,15 +53,74 @@ - (TKSignature *)signData:(NSData *)data
             usingKeyLevel:(Key_Level)keyLevel
                    reason:(NSString *)reason
                   onError:(OnError)onError {
+    if (!useLocalAuthentication) {
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    if (keyLevel >= Key_Level_Low) {
+        // We don't check DeviceOwnerAuthentication for Key_Level_Low
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    LAContext *context = [[LAContext alloc] init];
+    
+    if (![context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]) {
+        // If device don't support Device Owner Authentication, skip as success
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    __block TKSignature *signature = nil;
+    __block NSError *laError = nil;
+    
+    [context evaluatePolicy:LAPolicyDeviceOwnerAuthentication
+            localizedReason:reason
+                      reply:^(BOOL success, NSError *error) {
+                          // This is a backend thread
+                          if (success) {
+                              signature = [self createSignature:data usingKeyLevel:keyLevel];
+                          }
+                          else {
+                              laError = error;
+                          }
+                          dispatch_semaphore_signal(semaphore);
+                      }];
+    
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+    
+    if (signature == nil) {
+        // Handles error
+        TKLogError(@"Error signing data: %@", laError);
+        if ([laError.domain isEqual:@kLAErrorDomain] &&
+            (laError.code == kLAErrorUserCancel || laError.code == kLAErrorSystemCancel )) {
+            onError([NSError errorFromErrorCode:kTKErrorUserCancelled
+                                        details:TKLocalizedString(@"User_Cancelled_Authentication",
+                                                                  @"User cancelled authentication")
+                              encapsulatedError:laError]);
+        }
+        else {
+            onError([NSError errorFromErrorCode:kTKErrorUserInvalid
+                                        details:TKLocalizedString(@"User_Invalid_Authentication",
+                                                                  @"Invalid user for authentication")
+                              encapsulatedError:laError]);
+        }
+        return nil;
+    }
+    
+    return signature;
+}
+
+- (TKSignature *)createSignature:(NSData *)data
```

Comment:
Should this be private method?

Reply:
this method is not exposed in the header file. In objc, this is a private method already.

Reply:
We prepend _ to private methods and put them at the end of the file, by convention.